### PR TITLE
[Pattern] Add New Pattern Configurator UI

### PR DESCRIPTION
**Pull request by Abhishek-kumar09**
**Description**
Add new Configurator UI action type.
- [x] Optimizations: Workloads and traits are very large data(8.xx mb) of json when only istio adapter is running and it is fetched and parsed on every click to configure icons. Fixed it with one time fetching and parsing by memorizing the data and saving with context API approach.
- [x] Experimental new UI for patterns page
- [x] Toggler for list-view and individual view
- [x] Grouping the meshes with type and versions
- [x] Code refractor is needed to be done.
- [x] Have to fix problems with core definitions
- [x] Break it into more functional components 
- [x] Need more performance optimization
- [ ] Let’s try separating Kuberentes from the meshes.
- [x] Let’s take the easy path on displaying custom resources for each service mesh, by not using a toolbar full of icons, but a dropdown list (or some other selector). 
- [x] Human readable name to service name

Initial-video (draft):

https://user-images.githubusercontent.com/48255244/138336515-64bfe8a8-da4d-4441-a09a-3378552773bd.mp4

## latest Video: Iteration 2

https://www.youtube.com/watch?v=kQhd3gZCzLU

Fixes:

-Add performance optimization by using debouncer on state propagation
- Added common utilities functions for meshes
- One time data fetch and re-use saves a lot of network calls and unuseful parsing
- Configurator UI change, changes in design of RJSF components
- Fix crashing errors that is seen in the latest version of meshery when you open kubernetes service form in patterns

**Notes for Reviewers**

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by Abhishek-kumar09** [comment]
# latest Video: Iteration 2
https://www.youtube.com/watch?v=kQhd3gZCzLU

Fixes: 
1. Add performance optimization by using debouncer on state propagation
2. Added common utilities functions for meshes
3. One time data fetch and re-use saves a lot of network calls and unuseful parsing
4. Configurator UI change, changes in design of RJSF components
5. Fix crashing errors that is seen in the latest version of meshery when you open kubernetes service form in patterns

File: ui/components/ConfiguratorWrapper.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  }
+
+  useEffect(() => {
+    initialiser();
+  }, []);
+
+  return (
+    <SchemaContext.Provider value={{ workloadTraitSet, meshWorkloads }}>
+      {children}
+    </SchemaContext.Provider>
```
[the remark this answers is not in the document]
Comment (leecalcote) [reply ui/components/ConfiguratorWrapper.js:25]: @sonatype-lift ignore

**Comment by warunicorn19** [comment]
@Abhishek-kumar09 some merge conflicts to address.

File: ui/components/configuratorComponents/NameToIcon.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+/**
+ * nameToIcon returns icons for the rjsf form titles
+ *
+ * @param {string} name Tooltip name
+ * @param {string} color default value is primary
+ * @returns CustomIconButton
+ */
+export default function NameToIcon({ name, color = "#607D8B", ...other }) {
+
+  const CustomIcon = ({ Icon }) => <Icon style={{ color }}  {...other} />
```
[the remark this answers is not in the document]
Comment (leecalcote) [reply ui/components/configuratorComponents/NameToIcon.js:17]: @sonatype-lift ignore

**Comment by leecalcote** [comment]
We can split out Kubernetes later (or potentially not at all).

**Comment by Abhishek-kumar09** [comment]
> We can split out Kubernetes later (or potentially not at all).

Not adding kubernetes, sounds better to me especially for pattern configurator. Since it shows the active kubernetes resources in cluster and are not dependent on Meshery and may not be the subject to pattern configuration

**Comment by navendu-pottekkat** [comment]
@Abhishek-kumar09 Could you check the comments from the Lift bot?

File: ui/components/configuratorComponents/patternConfigurator.js [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+  function getTooltipTitleForIcons(schema) {
+    return schema?.workload?.metadata?.["display.ui.meshery.io/name"] || schema?.workload?.oam_definition?.spec?.metadata?.k8sKind || "N/A"
+  }
+
+  async function getPatternProps(schema) {
+    const refinedSchema = await getWorkloadTraitAndType(schema)
+    setActiveForm(refinedSchema)
+  }
+
+  if (!workloadTraitsSet) return <CircularProgress />
```
[the remark this answers is not in the document]
Comment (Abhishek-kumar09) [reply ui/components/configuratorComponents/patternConfigurator.js]: @sonatype-lift ignore

File: ui/components/configuratorComponents/NameToIcon.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+/**
+ * nameToIcon returns icons for the rjsf form titles
+ *
+ * @param {string} name Tooltip name
+ * @param {string} color default value is primary
+ * @returns CustomIconButton
+ */
+export default function NameToIcon({ name, color = "#607D8B", ...other }) {
+
+  const CustomIcon = ({ Icon }) => <Icon style={{ color }}  {...other} />;
```
[the remark this answers is not in the document]
Comment (Abhishek-kumar09) [reply ui/components/configuratorComponents/NameToIcon.js:17]: @sonatype-lift ignore

**Comment by Abhishek-kumar09** [comment]
Seems like the sonatype lift-bot is not configured for the React syntax. It is throwing errors following the rules of vanilla Js and eslint

**Comment by leecalcote** [comment]
Understanding that we'll wrap this up over the next few days, let's preemptively merge into an edge release and see how we're doing from there.

**Review by leecalcote — APPROVED** [review]

**Pull request merged**